expression/builtin_time.go

```diff
@@ -2569,159 +2632,372 @@ func isDuration(str string) bool {
 }
 
-// eval evals a builtinAddTimeSig.
+// evalTime evals a builtinAddTimeDateTimeNullSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_addtime
-func (b *builtinAddTimeSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+func (b *builtinAddTimeDateTimeNullSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	return types.ZeroDatetime, true, nil
+}
+
+type builtinAddDatetimeAndDurationSig struct {
+	baseTimeBuiltinFunc
+}
+
+// evalTime evals a builtinAddDatetimeAndDurationSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_addtime
+func (b *builtinAddDatetimeAndDurationSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	arg0, isNull, err := b.args[0].EvalTime(row, sc)
+	if isNull || err != nil {
+		return types.ZeroDatetime, isNull, errors.Trace(err)
+	}
+	arg1, isNull, err := b.args[1].EvalDuration(row, sc)
+	if isNull || err != nil {
+		return types.ZeroDatetime, isNull, errors.Trace(err)
+	}
+	result, err := arg0.Add(arg1)
+	return result, err != nil, errors.Trace(err)
+}
+
+type builtinAddDatetimeAndStringSig struct {
+	baseTimeBuiltinFunc
+}
+
+// evalTime evals a builtinAddDatetimeAndStringSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_addtime
+func (b *builtinAddDatetimeAndStringSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	arg0, isNull, err := b.args[0].EvalTime(row, sc)
+	if isNull || err != nil {
+		return types.ZeroDatetime, isNull, errors.Trace(err)
+	}
+	s, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return types.ZeroDatetime, isNull, errors.Trace(err)
+	}
+	if !isDuration(s) {
+		return types.ZeroDatetime, true, nil
+	}
+	arg1, err := types.ParseDuration(s, getFsp(s))
 	if err != nil {
-		return d, errors.Trace(err)
+		return types.ZeroDatetime, true, errors.Trace(err)
 	}
-	if args[0].IsNull() || args[1].IsNull() {
-		return
+	result, err := arg0.Add(arg1)
+	return result, err != nil, errors.Trace(err)
+}
+
+type builtinAddTimeDurationNullSig struct {
+	baseDurationBuiltinFunc
+}
+
+// evalDuration evals a builtinAddTimeDurationNullSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_addtime
+func (b *builtinAddTimeDurationNullSig) evalDuration(row []types.Datum) (types.Duration, bool, error) {
+	return types.ZeroDuration, true, nil
+}
+
+type builtinAddDurationAndDurationSig struct {
+	baseDurationBuiltinFunc
+}
+
+// evalDuration evals a builtinAddDurationAndDurationSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_addtime
+func (b *builtinAddDurationAndDurationSig) evalDuration(row []types.Datum) (types.Duration, bool, error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	arg0, isNull, err := b.args[0].EvalDuration(row, sc)
```

Comment:
can we change `arg0` to `lhs`, and `arg1` to `rhs` ?

Reply:
@zz-jason  `arg0` and `arg1` are of the same side of expression, I think `lhs` and `rhs` is not quite proper here.